### PR TITLE
Fix: Preventing adding duplicated task ID to taskOrder fix for #11385

### LINKED ISSUE #11385: challenge's tasksOrder array must not have duplicated task IDs
I have found a challenge that has duplicate task IDs in its tasksOrder object, as shown below. It's possible that this is causing https://github.com/HabitRPG/habitica/issues/11295 which is a highly inconvenient bug so I'm marking this as a high priority.

The expected behaviour is that every challenge task should appear once and only once in the challenge's tasksOrder object.

For anyone who wants to work on this, it might be best if you post here first with your intended fix to see if Habitica's staff approve of it (or look in the comments to see if staff have already described a preferred fix).

Loggly shows two errors during creation of the challenge (below), both saying `E11000 duplicate key error collection: habitica.tasks index: _id_ dup key: { : "[task id]" }`, and the task IDs in the errors match the duplicates. Thus it seems like a problem during the creation of those two challenge tasks caused the API to try to add the same task twice. The MongoDB index prevented the duplicate tasks being created but the IDs of the tasks were still inserted into the tasksOrder array (changing tasksOrder is done in a separate database update to the creation of the task).

Here is the challenge's tasksOrder object showing the duplicates:
```
{
    "_id": "703b2ea6-42a2-4b79-ae76-f6c871044aa7",
    "tasksOrder": {
        "habits": [],
        "dailys": [],
        "todos": [
            "d754dd2c-82b1-4bdc-b79d-9b568f79da7e",
            "dfdb1d51-d1ac-454c-8774-f3247333606a",
            "c32b7485-5c1d-41bd-9ab3-0acfd192dc0b",
            "f259e4a8-621a-4751-9b56-118ed5624118",
            "a987787d-1d55-4db4-abf8-b241f71fe6cc",
            "20493e76-925d-470f-935d-7a4ffe4d4572",
            "65b29b7c-11f6-4796-9117-5ab997312c8c",
            "61b11af7-4d59-420f-9e35-2d1001a3a6a5",
            "05a5bbac-19bf-4edc-8685-889cf9b0ef2e",
            "2bcae3ef-10e7-4ac8-adc5-5cd2eff8ee59",
            "a99ea1bf-6258-4c11-a448-29b1dbad3742",
            "e3766974-92bb-41b7-bf0d-a21b1b749eb1",
            "bd3f6dfe-ef63-44dd-ad46-79739bf06e8b",
            "1c1723d5-4e76-4c29-9044-31012130c99f",
            "1c1723d5-4e76-4c29-9044-31012130c99f", [duplicate]
            "9540175b-9cfc-4775-a366-8b0cc8bb0492",
            "fe972675-958b-4ee7-a209-f99ca779277c",
            "d62d4ada-9974-4d1e-a749-5a418045bbe6",
            "2d9e89d6-a7ca-4fce-b300-bbbee313f238",
            "2d9e89d6-a7ca-4fce-b300-bbbee313f238", [duplicate]
            "d69a8ebb-63ae-4186-811a-a6ed0b158de7",
            "40a90e58-f7c1-4835-a06b-8fc81835b66c",
            "82dcfb1e-7ba5-407b-af2b-5c90b23a5f98",
            "5f379bf1-ffe3-495e-8c02-a67fdfb9448b",
            "00f5a614-ccba-4754-be5b-1b4bafa607bc",
            "825fbd3f-d489-48c9-bc17-c38dcf94eb7e",
            "d991a6bf-65da-4ffc-9653-7de4d14bf1ea",
            "2d7a265d-2bea-4056-8950-626e97b63c61",
            "bfc7c2ba-4cd6-4149-8936-212ed8e02114",
            "af978892-c277-476f-abfe-4d1bad10035d",
            "a204f555-8435-459c-95a6-38f344b5acc3",
            "eb1c7576-ff1a-442f-b824-f75d1aa214a7",
            "4fed2a99-f4fd-4695-b3b5-9749fd473fca"
        ],
        "rewards": []
    },
}
```
As a note to other admins, I'm about to fix that specific challenge in the database so if you examine it you won't see that bug.

Here are the errors from Loggly:
```
{"method":"POST","originalUrl":"/api/v4/tasks/challenge/703b2ea6-42a2-4b79-ae76-f6c871044aa7","headers":{"host":"habitica.com","user-agent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:69.0) Gecko/20100101 Firefox/69.0","accept":"application/json, text/plain, */*","accept-language":"en-GB,en;q=0.5","accept-encoding":"gzip, deflate, br","x-client":"habitica-web","x-api-user":"2d8797a8-291d-4545-8bba-180d4e8701a6","x-user-timezoneoffset":"-60","content-type":"application/json;charset=utf-8","content-length":"961","referer":"https://habitica.com/challenges/703b2ea6-42a2-4b79-ae76-f6c871044aa7","x-cloud-trace-context":"6be0496c3a99e5bd8c3434fd0394e2ac/1871857669922801192","via":"1.1 google","x-forwarded-for":"86.24.196.156, 35.227.223.241","x-forwarded-proto":"https","connection":"Keep-Alive"},"body":{"0":{"type":"todo","text":"October 18th: No-Beard Day","_id":"1c1723d5-4e76-4c29-9044-31012130c99f","notes":"The first Saturday in September was Beard Day, and the 18th of October is its mirror-universe counterpart: No Beard Day. Beard Day is the one with the menacing little goatee, though, so I think we all know which is the evil twin here. If you're planning to celebrate Movember, today is an opportunity to get a fresh start before the hairy adventure ahead. If you have a magnificent beard you're very protective of, though, celebrate Chocolate Cupcake Day instead. Try not to get the icing in your face-fuzz.","tags":[],"priority":"1.0","reminders":[],"attribute":"str","collapseChecklist":false,"checklist":[],"date":"2019-10-18T18:28:00.000Z"}},"httpCode":500,"isHandledError":false,"fullError":{"driver":true,"name":"MongoError","index":0,"code":11000,
"errmsg":"E11000 duplicate key error collection: habitica.tasks index: _id_ dup key: { : \"1c1723d5-4e76-4c29-9044-31012130c99f\" }"},
"timestamp":"2019-09-23T18:28:43.103Z","level":"error","message":"MongoError: E11000 duplicate key error collection: habitica.tasks index: _id_ dup key: { : \"1c1723d5-4e76-4c29-9044-31012130c99f\" }\n    at Function.create (/usr/src/habitrpg/node_modules/mongodb-core/lib/error.js:43:12)\n    at toError (/usr/src/habitrpg/node_modules/mongodb/lib/utils.js:149:22)\n    at /usr/src/habitrpg/node_modules/mongodb/lib/operations/collection_ops.js:859:39\n    at handler (/usr/src/habitrpg/node_modules/mongodb-core/lib/topologies/replset.js:1155:22)\n    at fn (/usr/src/habitrpg/node_modules/@google-cloud/trace-agent/build/src/plugins/plugin-mongodb-core.js:77:13)\n    at /usr/src/habitrpg/node_modules/@google-cloud/trace-agent/build/src/cls/async-hooks.js:149:49\n    at AsyncHooksCLS.runWithContext (/usr/src/habitrpg/node_modules/@google-cloud/trace-agent/build/src/cls/async-hooks.js:127:20)\n    at contextWrapper (/usr/src/habitrpg/node_modules/@google-cloud/trace-agent/build/src/cls/async-hooks.js:149:25)\n    at /usr/src/habitrpg/node_modules/mongodb-core/lib/connection/pool.js:397:18\n    at processTicksAndRejections (internal/process/task_queues.js:75:11)"}
```

```
{"method":"POST","originalUrl":"/api/v4/tasks/challenge/703b2ea6-42a2-4b79-ae76-f6c871044aa7","headers":{"host":"habitica.com","user-agent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:69.0) Gecko/20100101 Firefox/69.0","accept":"application/json, text/plain, */*","accept-language":"en-GB,en;q=0.5","accept-encoding":"gzip, deflate, br","x-client":"habitica-web","x-api-user":"2d8797a8-291d-4545-8bba-180d4e8701a6","x-user-timezoneoffset":"-60","content-type":"application/json;charset=utf-8","content-length":"665","referer":"https://habitica.com/challenges/703b2ea6-42a2-4b79-ae76-f6c871044aa7","x-cloud-trace-context":"8d29f7ce535e39d2e1c6baf7ec93cd00/3893207610475662796","via":"1.1 google","x-forwarded-for":"86.24.196.156, 35.227.223.241","x-forwarded-proto":"https","connection":"Keep-Alive"},"body":{"0":{"type":"todo","text":"October 14th: Dessert Day","_id":"2d9e89d6-a7ca-4fce-b300-bbbee313f238","notes":"October 14th: Dessert Day is the perfect excuse to indulge in your favourite after-dinner treat. Eat it before dinner, or instead of dinner. If you've always bought it, why not experiment with making it yourself?","tags":[],"priority":"1.0","reminders":[],"attribute":"str","collapseChecklist":false,"checklist":[],"date":"2019-10-14T18:23:00.000Z"}},"httpCode":500,"isHandledError":false,"fullError":{"driver":true,"name":"MongoError","index":0,"code":11000,
"errmsg":"E11000 duplicate key error collection: habitica.tasks index: _id_ dup key: { : \"2d9e89d6-a7ca-4fce-b300-bbbee313f238\" }"},"
timestamp":"2019-09-23T18:24:21.502Z","level":"error","message":"MongoError: E11000 duplicate key error collection: habitica.tasks index: _id_ dup key: { : \"2d9e89d6-a7ca-4fce-b300-bbbee313f238\" }\n    at Function.create (/usr/src/habitrpg/node_modules/mongodb-core/lib/error.js:43:12)\n    at toError (/usr/src/habitrpg/node_modules/mongodb/lib/utils.js:149:22)\n    at /usr/src/habitrpg/node_modules/mongodb/lib/operations/collection_ops.js:859:39\n    at handler (/usr/src/habitrpg/node_modules/mongodb-core/lib/topologies/replset.js:1155:22)\n    at /usr/src/habitrpg/node_modules/mongodb-core/lib/connection/pool.js:397:18\n    at processTicksAndRejections (internal/process/task_queues.js:75:11)"}
```

### PR DESCRIPTION
**Pull request by cheng-shiqi**
[//]: # (Note: See http://habitica.fandom.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #11385 

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
In the `createTasks` function in the `habitica/website/server/libs/taskManager.js`, I added some checking code to prevent inserting duplicated task ID to the `taskOrder `objcect. 

So in a situation like API trying to create the same task twice(https://github.com/HabitRPG/habitica/issues/11385#issue-499858731), there would not be duplicated task IDs in the corresponding `taskOrder ` array.

_Before_, if I try to use API to create the same habit twice in a challenge, there would be a duplicated task ID in the taskOrder array:

snipaste_20191008_135846
snipaste_20191008_135927
_After_:

snipaste_20191008_142836
snipaste_20191008_142853

[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID: c57a0797-60a3-4811-9146-709a3312b823

### CODE REVIEW

**Comment by paglias** [comment]
Hi @cheng-shiqi and thanks for the PR. Would you be able to add a test to show that the fix is working? Or if that's not possible to show a list of API calls that result in duplicate IDs in the list?

**Comment by cheng-shiqi** [comment]
Hi @paglias  I tried but didn't managed to write a usable mocha test, that's why I test it manually:sweat_smile:. 
The test is expected to check if the taskOrder object have duplicate ID after two POST request containing same task ID, but this operation would get a `500 `error and terminate the test. I don't know how to make test work in this situation. Can anyone help with that?

Technically, I think any API(listed below) using the `createTasks `function from `abitica/website/server/libs/taskManager.js` , if it is called duplicatelly, will add duplicate task IDs to the `taskOrder `Object.

- `createUserTasks` in `habitica\website\server\controllers\api-v3\tasks.js`
- `createChallengeTasks` in `habitica\website\server\controllers\api-v3\tasks.js`
- `cloneChallenge` in `habitica\website\server\controllers\api-v3\challenges.js`
- `createGroupTasks` in `habitica\website\server\controllers\api-v3\tasks\groups.js`

please see https://github.com/HabitRPG/habitica/issues/11385#issue-499858731, there is a case @Alys found, in which `createChallengeTasks` was called twice and cause duplicate task IDs.

**Comment by cheng-shiqi** [comment]
Maybe I could set this to be ready for review. 

Hello, @paglias and @Alys , is the information above enough for reviewing this fix? Please leave comments if you have any ideas revising or improving this PR.

**Comment by paglias** [comment]
@cheng-shiqi I haven’t had time to review PRs last week, I’ll try this one. My concern is that the bug is caused by a race condition so I don’t think the check will help much but the only way to verify that is with a test

**Comment by paglias** [comment]
Actually it could work. I thought it wouldn’t because the dB prevents saving multiple tasks with the same Id but given that the update to save the tasks ids in the challenge is executed before the task is created it wouldn’t catch that.

Could you use an array instead of a Set? You can use .includes() on arrays as well

I’ll see to write a test for this but I can’t assure it’ll be done this week

File: website/server/libs/taskManager.js [hunk 1]
```diff
@@ -113,7 +113,7 @@ export async function createTasks (req, res, options = {}) {
 
     // Otherwise update the user/challenge/group
     if (!taskOrderToAdd[`${taskType}s`]) taskOrderToAdd[`${taskType}s`] = [];
-    taskOrderToAdd[`${taskType}s`].unshift(newTask._id);
+    if (!owner.tasksOrder[`${taskType}s`].includes(newTask._id) & !taskOrderToAdd[`${taskType}s`].includes(newTask._id)) taskOrderToAdd[`${taskType}s`].unshift(newTask._id);
```
Comment (paglias) [inline website/server/libs/taskManager.js]: I think this is missing a second &, but I don't think the second check is necessary (the one for taskOrderToAdd[`${taskType}s`])

**Review by paglias — CHANGES_REQUESTED** [review]

File: website/server/libs/taskManager.js (hunk 1 above)
Comment (cheng-shiqi) [reply website/server/libs/taskManager.js]: Sure, Maybe I was fussing over it, I've deleted it.

**Review by paglias — APPROVED** [review]

**Comment by paglias** [comment]
Fixed some conflicts generated by changes in the client side code

**Comment by paglias** [comment]
thanks @cheng-shiqi and sorry for the delay, I've marked this as ready and will merge as soon as possible

**Comment by paglias** [comment]
Thanks @cheng-shiqi ! The code will undergo some final testing and will then go live at the end of next week. I've awarded your second tier :)

**Pull request merged**